internal/codegen/stencil.go

```diff
@@ -99,9 +117,17 @@ func (s *Stencil) Render(ctx context.Context, log logrus.FieldLogger) ([]*Templa
 		}
 	}
 
-	// Now we render each file
+	// Render the first pass, this is used to populate shared data
```

Comment:
Instead of two fixed passes and forcibly re-rendering every template, I suggest a slightly different technique:

With each module hook, track the templates that read from it (hook subscribers).
When  a hook is modified (i.e. AddModuleHook is called but with different contents than before), loop through all subscribers and mark them `dirty`.

When a pass of rendering is done, repeat it only for dirty templates.     We can support only two passes and if there are dirty templates still abort with error that templates are abusing hooks incorrectly.

Reply:
I figured you might say this, I'll get it on it :D

Reply:
If you want a quick & dirty version, you can do two hardcoded passes, with the second pass only going through dirty templates.  And you can simplify dirty template tracking by only marking a template dirty if it called any code to fetch any hook (no need to track full subscribers list)

Reply:
Going to implement this in a future PR -- I need to get some templates in (Monday at the latest, just need to get this stencil version out and the other branch)